File: app/src/androidTest/java/org/andstatus/app/timeline/SharingMediaToThisAppTest.java
package org.andstatus.app.timeline;

import android.content.Intent;
import android.net.Uri;
import android.provider.BaseColumns;
import android.support.test.espresso.action.TypeTextAction;
import android.view.View;
import android.widget.TextView;

import org.andstatus.app.ActivityTestHelper;
import org.andstatus.app.R;
import org.andstatus.app.account.AccountSelector;
import org.andstatus.app.account.MyAccount;
import org.andstatus.app.activity.ActivityViewItem;
import org.andstatus.app.context.MyContextHolder;
import org.andstatus.app.context.TestSuite;
import org.andstatus.app.data.DownloadData;
import org.andstatus.app.data.DownloadStatus;
import org.andstatus.app.data.MyContentType;
import org.andstatus.app.data.MyQuery;
import org.andstatus.app.database.table.NoteTable;
import org.andstatus.app.service.MyServiceTestHelper;
import org.andstatus.app.util.MyLog;
import org.junit.After;
import org.junit.Test;

import java.util.Arrays;

import static android.support.test.espresso.Espresso.onView;
import static android.support.test.espresso.matcher.ViewMatchers.withId;
import static org.andstatus.app.context.DemoData.demoData;
import static org.junit.Assert.assertEquals;
import static org.junit.Assert.assertTrue;

public class SharingMediaToThisAppTest extends TimelineActivityTest<ActivityViewItem> {
    private MyServiceTestHelper mService;
    private MyAccount ma;

    @Override
    protected Intent getActivityIntent() {
        MyLog.i(this, "setUp started");
        TestSuite.initializeWithData(this);

        mService = new MyServiceTestHelper();
        mService.setUp(demoData.gnusocialTestAccountName);
        ma = demoData.getMyAccount(demoData.gnusocialTestAccountName);
        assertTrue(ma.isValid());
        MyContextHolder.get().accounts().setCurrentAccount(ma);

        Intent intent = new Intent(Intent.ACTION_SEND);
        intent.setType("image/png");
        Uri mediaUri = demoData.localImageTestUri2;
        assertTrue(mediaUri != null);
        MyLog.i(this, "setUp ended");
        intent.putExtra(Intent.EXTRA_STREAM, mediaUri);
        return intent;
    }

    @After
    public void tearDown() throws Exception {
        mService.tearDown();
    }

    @Test
    public void testSharingMediaToThisApp() throws InterruptedException {
        final String method = "testSharingMediaToThisApp";
        ListActivityTestHelper<TimelineActivity> listActivityTestHelper =
                ListActivityTestHelper.newForSelectorDialog(getActivity(), AccountSelector.getDialogTag());
        listActivityTestHelper.selectIdFromSelectorDialog(method, ma.getActorId());

        View editorView = getActivity().findViewById(R.id.note_editor);
        ActivityTestHelper.waitViewVisible(method, editorView);
        TextView details = editorView.findViewById(R.id.noteEditDetails);
        String textToFind = MyContextHolder.get().context().getText(R.string.label_with_media).toString();
        ActivityTestHelper.waitTextInAView(method, details, textToFind);

        String body = "Test note with a shared image " + demoData.testRunUid;
        TestSuite.waitForIdleSync();
        onView(withId(R.id.noteBodyEditText)).perform(new TypeTextAction(body));
        TestSuite.waitForIdleSync();

        mService.serviceStopped = false;
        ActivityTestHelper<TimelineActivity> helper = new ActivityTestHelper<>(getActivity());
        helper.clickMenuItem(method, R.id.noteSendButton);
        ActivityTestHelper.waitViewInvisible(method, editorView);

        mService.waitForServiceStopped(false);

        String message = "Data was posted " + mService.getHttp().getPostedCounter() + " times; "
                + Arrays.toString(mService.getHttp().getResults().toArray());
        MyLog.v(this, method + "; " + message);
        assertTrue(message, mService.getHttp().getPostedCounter() > 0);
        assertTrue(message, mService.getHttp().substring2PostedPath("statuses/update").length() > 0);

        String condition = "BODY='" + body + "'";
        long unsentMsgId = MyQuery.conditionToLongColumnValue(NoteTable.TABLE_NAME, BaseColumns._ID, condition);
        assertTrue("Unsent note found: " + condition, unsentMsgId != 0);
        assertEquals("Status of unsent note", DownloadStatus.SENDING, DownloadStatus.load(
                MyQuery.noteIdToLongColumnValue(NoteTable.NOTE_STATUS, unsentMsgId)));

        DownloadData dd = DownloadData.getSingleForNote(unsentMsgId,
                MyContentType.IMAGE, null);
        MyLog.v(this, method + "; " + dd);
        assertEquals("Image URI stored", demoData.localImageTestUri2, dd.getUri());
        assertEquals("Loaded '" + dd.getUri() + "'; " + dd, DownloadStatus.LOADED, dd.getStatus());
    }
}

File: app/src/androidTest/kotlin/org/andstatus/app/timeline/SharingMediaToThisAppTest.kt
package org.andstatus.app.timeline

import android.content.Intent
import android.net.Uri
import android.provider.BaseColumns
import android.view.View
import android.widget.TextView
import androidx.test.espresso.Espresso
import androidx.test.espresso.action.ReplaceTextAction
import androidx.test.espresso.matcher.ViewMatchers
import org.andstatus.app.ActivityTestHelper
import org.andstatus.app.R
import org.andstatus.app.account.MyAccount
import org.andstatus.app.activity.ActivityViewItem
import org.andstatus.app.context.DemoData
import org.andstatus.app.context.MyContext
import org.andstatus.app.context.TestSuite
import org.andstatus.app.data.DownloadData
import org.andstatus.app.data.DownloadStatus
import org.andstatus.app.data.MyQuery
import org.andstatus.app.database.table.NoteTable
import org.andstatus.app.service.MyServiceTestHelper
import org.andstatus.app.util.EspressoUtils
import org.andstatus.app.util.MyLog
import org.andstatus.app.view.SelectorDialog
import org.junit.After
import org.junit.Assert
import org.junit.Test
import kotlin.properties.Delegates

class SharingMediaToThisAppTest : TimelineActivityTest<ActivityViewItem>() {
    private val myContext: MyContext = TestSuite.initializeWithAccounts(this)
    private var mService: MyServiceTestHelper by Delegates.notNull()
    private var ma: MyAccount = MyAccount.EMPTY

    override fun getActivityIntent(): Intent {
        MyLog.i(this, "setUp started")
        mService = MyServiceTestHelper()
        mService.setUp(DemoData.demoData.gnusocialTestAccountName)
        ma = DemoData.demoData.getGnuSocialAccount()
        Assert.assertTrue(ma.isValid)
        myContext.accounts.setCurrentAccount(ma)
        val intent = Intent(Intent.ACTION_SEND)
        intent.type = "image/png"
        val mediaUri: Uri = DemoData.demoData.localImageTestUri2
        Assert.assertNotNull(mediaUri)
        intent.putExtra(Intent.EXTRA_STREAM, mediaUri)
        MyLog.i(this, "setUp ended")
        return intent
    }

    @After
    fun tearDown() {
        mService.tearDown()
    }

    @Test
    fun testSharingMediaToThisApp() {
        val method = "testSharingMediaToThisApp"
        TestSuite.waitForListLoaded(activity, 4)

        val listActivityTestHelper: ListActivityTestHelper<TimelineActivity<*>> =
            ListActivityTestHelper.newForSelectorDialog(activity, SelectorDialog.Companion.dialogTag)
        listActivityTestHelper.selectIdFromSelectorDialog(method, ma.actorId)
        val editorView = activity.findViewById<View?>(R.id.note_editor)
        ActivityTestHelper.waitViewVisible(method, editorView)
        val details = editorView.findViewById<TextView?>(R.id.noteEditDetails)
        val textToFind: String = myContext.context.getText(R.string.label_with_media).toString()
        ActivityTestHelper.waitTextInAView(method, details, textToFind)
        EspressoUtils.waitForIdleSync()
        val content = "Test note with a shared image " + DemoData.demoData.testRunUid
        Espresso.onView(ViewMatchers.withId(R.id.noteBodyEditText)).perform(ReplaceTextAction(content))
        EspressoUtils.waitForIdleSync()

        ActivityTestHelper.clickSendButton(method, activity)
        val found = mService.waitForCondition {
            getHttp().substring2PostedPath("statuses/update").isNotEmpty()
        }
        val message = ("Data was posted " + mService.getHttp().getPostedCounter() + " times; " +
            mService.getHttp().getResults().toTypedArray().contentToString())
        MyLog.v(this, "$method; $message")
        Assert.assertTrue(message, mService.getHttp().getPostedCounter() > 0)
        Assert.assertTrue(message, found)

        val condition = NoteTable.CONTENT + "='" + content + "'"
        val unsentMsgId = MyQuery.conditionToLongColumnValue(NoteTable.TABLE_NAME, BaseColumns._ID, condition)
        Assert.assertTrue("Unsent note found: $condition", unsentMsgId != 0L)
        Assert.assertEquals(
            "Status of unsent note", DownloadStatus.SENDING, DownloadStatus.Companion.load(
                MyQuery.noteIdToLongColumnValue(NoteTable.NOTE_STATUS, unsentMsgId)
            )
        )
        val dd: DownloadData = DownloadData.Companion.getSingleAttachment(unsentMsgId)
        MyLog.v(this, "$method; $dd")
        Assert.assertEquals("Image URI stored", DemoData.demoData.localImageTestUri2, dd.getUri())
        Assert.assertEquals("Loaded '" + dd.getUri() + "'; " + dd, DownloadStatus.LOADED, dd.getStatus())
    }
}
